assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx

```diff
@@ -0,0 +1,257 @@
+ * This ensures that when the terms are displayed in the `FormTokenField`, each term name remains unique.
+ *
+ * An illustrative example of how termIdToNameMap might look is as follows:
+ * {
+ *    "19": "Accessories",
+ *    "37": "category1 - category1",
+ *    "38": "category1 - category1-clothing",
+ *    "39": "category1 - category1-clothing-2",
+ *    "16": "Clothing",
+ *    "21": "Decor"
+ * }
+ * In the example above, "category1" is a duplicated term name, so the term's slug is appended for distinction.
+ *
+ * termNameToIdMap is the inverse of termIdToNameMap, mapping term names back to their respective IDs.
+ */
+const useTermMaps = (
+	taxonomy: Taxonomy
+): {
+	termIdToNameMap: Map< number, string >;
+	termNameToIdMap: Map< string, number >;
+	isResolving: boolean;
+} => {
+	// Fetch all terms for the given taxonomy.
+	const { records: allTerms, isResolving: isResolvingAllTerms } =
+		useEntityRecords< Term[] >( 'taxonomy', taxonomy.slug, {
+			...BASE_QUERY_ARGS,
+		} );
+
+	// Memoize the result to avoid re-renders.
+	return useMemo( () => {
+		const termIdToNameMap = new Map< number, string >();
+		const termNameToIdMap = new Map< string, number >();
+
+		if ( ! allTerms )
+			return {
+				termIdToNameMap,
+				termNameToIdMap,
+				isResolving: isResolvingAllTerms,
+			};
+
+		// Count the number of times a term name appears.
+		const nameCountMap = createNameCountMap( allTerms );
+
+		// Create the map with term ids as keys and term names as values.
+		for ( const term of allTerms ) {
+			const termId = term.id;
+			const termName = term.name;
+			const name = generateUniqueName(
+				termName,
+				term.slug,
+				nameCountMap
+			);
+			termIdToNameMap.set( termId, name );
+			termNameToIdMap.set( name, termId );
+		}
+		return {
+			termIdToNameMap,
+			termNameToIdMap,
+			isResolving: isResolvingAllTerms,
+		};
+	}, [ allTerms, isResolvingAllTerms ] );
+};
+
+const TaxonomyItem = ( { taxonomy, termIds, onChange }: TaxonomyItemProps ) => {
+	const [ search, setSearch ] = useState( '' );
+	const suggestionsRef = useRef< string[] >( EMPTY_ARRAY );
+	const currentValueRef = useRef<
+		{
+			id: number;
+			value: string;
+		}[]
+	>( EMPTY_ARRAY );
+
+	// Search is debounced to limit the number of API calls as the user types
+	const debouncedSearch = useDebounce( setSearch, 250 );
+
+	const {
+		termIdToNameMap,
+		termNameToIdMap,
+		isResolving: isResolvingTermMaps,
+	} = useTermMaps( taxonomy );
+
+	// Fetch the terms based on the search query.
+	const { records: searchResults, hasResolved: searchHasResolved } =
+		useEntityRecords( 'taxonomy', taxonomy.slug, {
+			...BASE_QUERY_ARGS,
+			search,
+			orderby: 'name',
+			exclude: termIds,
+			per_page: 20,
+		} );
+
+	suggestionsRef.current = useMemo( () => {
+		if ( ! searchHasResolved ) return suggestionsRef.current;
+
+		const newSuggestions = searchResults.map( ( searchResult: Term ) =>
+			termIdToNameMap.get( searchResult.id )
```

Comment:
On my end, this causes a JS error when `Uncategorized` is one of the results, as it doesn't seem to be present in `termIdToNameMap` for some reason. I don't really know why. Could we add a fallback to `searchResult.name` just in case?

```suggestion
			termIdToNameMap.get( searchResult.id ) || searchResult.name
```

Reply:
I can think of only one case when this might happen, i.e. when we are editing Product Collection block in a post and in a new tab we add a new term. In that case, new term won't be available in termIdToNameMap until user refreshes the page. 

Can you think of any other case when this might happen? 

I am adding a fallback for such scenarios. 🙂

Reply:
Hmmm, I'm not sure what was going on. Tested on a different site and I can't reproduce it there. Let's leave it as-is and I will open an issue if I can reproduce again. Sorry for the confusion. :sweat: